architecture/ambient/lifecycle.md

```diff
@@ -0,0 +1,136 @@
+# Ztunnel Lifecyle
+
+This document describes the lifecycle of various aspects of Ztunnel, and its relationship to workloads and the CNI.
+
+At a high level, our goal is to provide complete connectivity to a workload throughout its entire lifetime.
+Failing to do so can be an availability risk (if we deny traffic that should succeed) or a security risk (if we allow traffic that should be denied).
+
+## High level overview
+
+At a high level, the relevant components look as such:
+
+```mermaid
+flowchart TD
+    CNIP["CNI Plugin"]
+    CNI["CNI Agent"]
+    Z["Ztunnel"]
+    Pods["Pods"]
+    CNI --Programs--> Z
+    CNI --Watches--> Pods
+    Z --Runs Within--> Pods
+    CNIP --Calls--> CNI
+```
+
+The CNI Plugin is a binary installed as a [CNI plugin](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/) on the node.
```

Comment:
Should we just put some/all of this stuff in https://istio.io/latest/docs/ambient/architecture/traffic-redirection/ ?

I feel like most of the generic `lifecycle` stuff and calling out of specific security considerations and our mitigations of them would be useful to have there.

e.g. "It is security-critical to handle pod startup gaps for this reason, and we do in this way" probably belongs in `istio.io` docs.

I don't want to document the CNI node agent in too many disparate places, we already have

- https://istio.io/latest/docs/ambient/architecture/traffic-redirection/ (ambient-specific generic traffic capture architecture)

- https://istio.io/latest/docs/setup/additional-setup/cni/ (dataplane-mode agnostic install guide)
  
cc @craigbox for thoughts.

Reply:
I agree some probably should be in public docs but there is serious value to having developer docs like this that is not tied up in public docs personas

Reply:
I would go one step further - public docs should not be concerned with implementation details (that may change or depend on a particular implementation).

What is relevant for public docs is "The ambient L4 layer guarantees that network is secure ... from start to end"

Reply:
Implementation details should not be covered in public docs - agreed.

But I don't think "K8S-specific pod lifecycle security considerations" are component-level implementation details. That's the whole reason/justification we use a CNI plugin, and have repair for sidecars, etc etc. It's worth explaining that at a high level.

We have already had people ask about these basic architectural justifications, is why I'm suggesting we have something in public docs.

Maybe what's already there is sufficient - that's fine. Wanna make sure it is.

Reply:
> I would go one step further - public docs should not be concerned with implementation details (that may change or depend on a particular implementation).

I don't think we can get away with that, since the entire reason we have a node agent with a CNI plugin has to do with K8S pod lifecycle implementation details.

Reply:
Implementation details in docs are good for a few reasons:
* You did something great and need to prove it's real (we provide secure networking with X properties,and here is the details if you really want)
* You did something terrible and need to explain it (sidecar init has issues so we hacked on a repair controller)
* Understanding details is useful for operations (arguably, this is the same as "you did something terrible", because making users care about low level details means we did a bad job making it operational)

I think we have a bit of each here (hopefully less terrible stuff)

Reply:
Yeah - I'm more concerned that we have those implementation details here _and_ elsewhere too, and I'd rather be parsimonious about duplicating those details in different spots.

Putting them here is fine - if we do a pass to remove those same details from the existing spots, if and where they exist.

Reply:
I agree that this is content for two separate audiences and it's OK to have some overlap.
Once this is in, we can easily mine the best bits and see what should go into the user-facing documentation.

Reply:
Then let's at least link to this from the `istio-cni` README then, since it already has a TODO stub for it.

Reply:
(and scrub a lot of the content from that README too, actually - it mostly restates a subset of what's here now)